common/channel/src/libra_channel.rs

```diff
@@ -42,24 +43,46 @@ struct SharedState<T> {
     /// consumption by the Receiver and then the executor/scheduler will wake up
     /// the Receiver task to process the next item.
     waker: Option<Waker>,
+
+    /// A boolean which tracks whether the receiver has dropped
+    receiver_dropped: bool,
+    /// A boolean which tracks whether the sender has dropped
+    sender_dropped: bool,
+    /// A boolean which tracks whether the stream has terminated
+    /// A stream is considered terminated when sender has dropped
+    /// and we have drained everything inside our internal queue
+    stream_terminated: bool,
 }
 
 /// The sending end of the libra_channel.
-#[derive(Clone)]
 pub struct Sender<T> {
     shared_state: Arc<Mutex<SharedState<T>>>,
```

Comment:
random q: why do we need `MessageQueue` to be a trait? it seems simpler to just have `<K: Eq + Hash, V>` type parameters. the `Sender`/`Receiver` designs are also tightly coupled to the underlying `PerValidatorQueue` design imo; it is conceivable that they could be implemented without a `Mutex` etc or using many tokio channels which would motivate a different API anyway.

Reply:
So `MessageQueue` is the internal queue data structure in this channel. When something is `pushed` into this channel, we store it in `MessageQueue`. It is a trait to decouple the channel   from the internal storage/queueing mechanism. Without this trait, the channel implementation would require a concrete implementation of an internal queueing mechanism for messages.

> Sender/Receiver designs are also tightly coupled to the underlying PerValidatorQueue 
True. The current Mutex-based API is definitely not the most optimal design and we can iterate on improving this in the future.

Reply:
1. I guess I don't see what other use cases we have for substituting the `PerValidatorQueue` with... something else? while still maintaining the exact same API above? Maybe I am just missing something, in which I apologize for being slow :).
2. small nit: I definitely did a bit of a double take when I saw, e.g., `Sender<T>` where `T: MessageQueue`, since in every other queue implementation I've read, `T` is the message type. For k/v-type things, I've always seend `K, V` as the type parameters.

Reply:
+ For example, one of my concerns is that, for example, `libra_channel::Receiver<PerValidatorQueue<ProposalMsg<T>>>` is a very long type signature, when it could be something like `fair_channel::Receiver<ProposalMsg<T>>`, where the key and inner queue are just fixed and not needed to be specified. of course, `fair_channel::Receiver<T>` could just be a type alias for `libra_channel::Receiver<PerValidatorQueue<T>>`.

Reply:
That makes sense. However I don't want to make those changes in this PR. I'll open another PR for getting rid of `MessageTrait` and making `PerValidatorQueue` a part of `libra_channel`.